### PR TITLE
add ldap escaping for RFC 2253

### PR DESCRIPTION
**Pull request by loomy**
fixes a bug for ldap auth_binds for usernames including ldap special chars.

see https://www.mail-archive.com/dovecot@dovecot.org/msg66721.html

### CODE REVIEW

File: src/auth/db-ldap.c [hunk 1]
```diff
@@ -1435,6 +1434,35 @@ db_ldap_value_get_var_expand_table(struct auth_request *auth_request,
 	return table;
 }
 
+/* rfc2253 escaping */
+#define IS_LDAPDN_ESCAPED_CHAR(c) \
+	((c) == '"' || (c) == '+' || (c) == ',' || (c) == '\\' || (c) == '<' || (c) == '>' || (c) == ';')
```
Comment (nicolas-grekas) [inline src/auth/db-ldap.c]: what about `=`, `#`, `\r`, and leading trailing white spaces?

File: src/auth/passdb-ldap.c [hunk 2]
```diff
@@ -198,9 +198,8 @@ static void ldap_auth_bind(struct ldap_connection *conn,
 				     auth_request);
 		return;
 	}
-
-	brequest->request.callback = ldap_auth_bind_callback;
-	db_ldap_request(conn, &brequest->request);
+        brequest->request.callback = ldap_auth_bind_callback;
+        db_ldap_request(conn, &brequest->request);
```
Comment (nicolas-grekas) [inline src/auth/passdb-ldap.c]: spaces vs tabs here

File: src/auth/db-ldap.c [hunk 3]
```diff
@@ -304,7 +304,6 @@ static int db_ldap_request_bind(struct ldap_connection *conn,
 	i_assert(conn->conn_state == LDAP_CONN_STATE_BOUND_AUTH ||
 		 conn->conn_state == LDAP_CONN_STATE_BOUND_DEFAULT);
 	i_assert(conn->pending_count == 0);
-
```
Comment (cmouse) [inline src/auth/db-ldap.c]: unrelated whitespace change

File: src/auth/db-ldap.h [hunk 4]
```diff
@@ -199,6 +199,8 @@ void db_ldap_enable_input(struct ldap_connection *conn, bool enable);
 
 const char *ldap_escape(const char *str,
 			const struct auth_request *auth_request);
+const char *ldapdn_escape(const char *str,
```
Comment (cmouse) [inline src/auth/db-ldap.h]: why not just extend ldap_escape?

**Comment by loomy** [comment]
you are both right. 

the rfc is not fully implemented and space tab is messed too.

just wanted a quick check, if you are interested in this bug, as there wasnt any reaction on the mailinglist for this problem.

i could implement this in the ldap_escape function. but this will last some time, as I am on holiday from friday and needed this fix for a customer

so I will close this one and create a new pull request, when finished

**Comment by cmouse** [comment]
Please do, thank you for your effort!
